prompting/tools/datasets/arxiv.py

```diff
@@ -0,0 +1,227 @@
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+# DEALINGS IN THE SOFTWARE.
+
+import re
+import sys
+import fitz
+import random
+import datetime
+from dataclasses import dataclass
+import bittensor as bt
+import arxiv
+from typing import Dict, Union, List, Tuple
+from queue import Queue, Full, Empty
+from functools import lru_cache
+from .base import Dataset
+from ..selector import Selector
+
+
+# Create a queue called CACHED_ARTICLES to store arxiv articles that have been fetched
+CACHED_ARTICLES = Queue(maxsize=300)
+
+client = arxiv.Client()
+
+
+@dataclass
+class ArxivPaper:
+
+    # Define common section headers to search for
+    section_headers = [
+        r"(?i)\babstract\b",
+        r"(?i)\bintroduction\b",
+        r"(?i)\bmethods\b",
+        r"(?i)\bmethodology\b",
+        r"(?i)\bresults\b",
+        r"(?i)\bdiscussion\b",
+        r"(?i)\bconclusion\b",
+        r"(?i)\breferences\b",
+    ]
+
+    def __init__(self, paper, text):
+        self.paper = paper
+        self.text = text
+
+        self.parse(text)
+
+    def parse(self, text):
+        # Find sections using regular expressions
+        sections = {}
+        for header in self.section_headers:
+            match = re.search(header, text)
+            if match:
+                sections[match.group().lower()] = match.start()
+
+        # Sort sections by their position in the text
+        sorted_sections = sorted(sections.items(), key=lambda x: x[1])
+
+        # Extract sections
+        paper_sections = {}
+        for i in range(len(sorted_sections)):
+            section_name = sorted_sections[i][0]
+            section_start = sorted_sections[i][1]
+            section_end = (
+                sorted_sections[i + 1][1] if i + 1 < len(sorted_sections) else len(text)
+            )
+            paper_sections[section_name] = text[section_start:section_end].strip(
+                f" \n{section_name}"
+            )
+
+        self.sections = paper_sections
+
+
+# speed up page loading
+@lru_cache(maxsize=1000)
+def _get_text(paper, seed=None) -> arxiv.Result:
+    """Cached Arxiv paper loading."""
+    try:
+        pdf_filename = paper.download_pdf()
+        # Read the PDF file
+        with fitz.open(pdf_filename) as pdf_document:
+            num_pages = pdf_document.page_count
+            text = ""
+            for page_num in range(num_pages):
+                page = pdf_document.load_page(page_num)
+                text += page.get_text()
+
+        # If paper is terribly formatted, return None
+        if text.count("\n") / len(text.split()) > 0.3:
+            bt.logging.warning(f"Paper {paper!r} is poorly formatted. Rejecting.")
+            return None
+
+        # Returns the pdf text
+        return text
```

Comment:
Are we deleting PDFs afterwards, or would it be better to use `TemporaryDirectory`? Also list would be more efficient than string concat.
Not tested:

```suggestion
        with tempfile.TemporaryDirectory() as tempdir:
            pdf_filename = paper.download_pdf(dir=tempdir)
            
            # Read the PDF file
            text: list[str] = []
            with fitz.open(pdf_filename) as pdf_document:
                for page in pdf_document:
                    text.append(page.get_text())
            
            text = "".join(text)

            # If paper is terribly formatted, return None
            if text.count("\n") / len(text.split()) > 0.3:
                bt.logging.warning(f"Paper {paper!r} is poorly formatted. Rejecting.")
                return None

            # Return the PDF text
            return text
```